### ISSUE TITLE
Inconsistent Handling of Wave Function Phase 

### ISSUE THREAD

**Issue by rcclay** [QMCPACK/qmcpack#1030]
I contend that this is a bug, but I'll throw this to the group before I start changing things.  Perhaps others are more aware of the ramifications.  It appears as though the real and complex builds handle the wave function phase information in different and contradictory manners.  Here's an example from src/QMCWavefunctions/TrialWavefunction.cpp::ratio(ParticleSet& P, int iat) (line 314).  

```c++
#if defined(QMC_COMPLEX)
  //return std::exp(evaluateLogAndPhase(r,PhaseValue));
  RealType logr=evaluateLogAndPhase(r,PhaseDiff);
  return std::exp(logr);
#else
  if (r<0)
    PhaseDiff=M_PI;
  //     else PhaseDiff=0.0;
  return r;
#endif
```
Basically, the complex build converts `ValueType r` to a real magnitude (positive semi-definite) and real phase.  In contrast, the real build notes that if `r` is negative, it sets the phase to π, but then doesn't do anything to change `r` to a positive number.

**Comment by rcclay** [comment]
The ratio information is correct, provided you know whether you are in an #ifdef QMC_COMPLEX block or not, and you know what the returned value of ratio() means in each case.  If QMC_COMPLEX is true, its expected that you'll take the returned "ratio", then get the phase, then reconstruct the ratio yourself, whereas if QMC_COMPLEX is false, then you can take the returned value with no more work.  This seems unsettling to me.  At the very least, PhaseDiff should be set to zero in the real case, since adding a phase of Pi only makes sense if you're  not allowing the ratio to be negative, which we are in the real case.

**Comment by rcclay** [comment]
Calling it a bug was too strong.

**Comment by ye-luo** [comment]
In drivers, ratio^2 is used and the sign doesn't matter. In NLPP, the sign of ratio is needed.

**Comment by ye-luo** [comment]
Currently ifdef is floating everywhere in NLPP and CUDA drivers. A clean way to hide complex/real is preferred.

**Comment by ye-luo** [comment]
Fix by  #2048
